File: Artist/app/src/main/java/com/example/artist/ui/topartistslisting/TopArtistsFragment.java
package com.example.artist.ui.topartistslisting;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.artist.R;
import com.example.artist.adapters.TopArtistsAdapter;
import com.example.artist.models.Artist;
import com.example.artist.ui.BaseFragment;
import com.example.artist.ui.DetailsActivity;
import com.example.artist.ui.topartistslisting.di.DaggerTopArtistsComponent;
import com.example.artist.ui.topartistslisting.di.TopArtistsModule;
import com.example.artist.utils.Constants;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;


public class TopArtistsFragment extends BaseFragment implements TopArtistsView {
    @BindView(R.id.rclr_artists)
    RecyclerView artistsRecyclerView;
    @BindView(R.id.prgrs_main)
    ProgressBar mainProgressBar;
    private OnFragmentInteractionListener mListener;
    @BindView(R.id.empty_layout)
    View emptyLayout;
    @Inject
    TopArtistsPresenter mPresenter;
    TopArtistsAdapter mAdapter;

    public TopArtistsFragment() {
        // Required empty public constructor
    }

    @Override
    protected void searchUserName(String userName) {
        if (mAdapter != null) {
            mAdapter.clearDataset();
        }
        mPresenter.getUserTopArtists(userName, Constants.TOP_ITEMS_LIMIT, Constants.API_KEY);

    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_artists, container, false);
        ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        DaggerTopArtistsComponent.builder().topArtistsModule(new TopArtistsModule(this)).build().inject(this);
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mPresenter.getUserTopArtists(Constants.DEFAULT_LASTFM_USER, Constants.TOP_ITEMS_LIMIT, Constants.API_KEY);
    }


    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnFragmentInteractionListener) {
            mListener = (OnFragmentInteractionListener) context;
        } else {
            throw new RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener");
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mListener = null;
    }

    @Override
    public void showProgress() {
        mainProgressBar.setVisibility(View.VISIBLE);
    }

    @Override
    public void hideProgress() {
        mainProgressBar.setVisibility(View.GONE);
    }

    @Override
    public void updateData(List<Artist> topArtists) {
        if (mAdapter == null) {
            mAdapter = new TopArtistsAdapter(topArtists, getContext(), onArtistclickedListener);
            LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext(), LinearLayoutManager.VERTICAL, false);
            artistsRecyclerView.setLayoutManager(linearLayoutManager);
            artistsRecyclerView.setAdapter(mAdapter);
            mAdapter.notifyDataSetChanged();
        } else {
            mAdapter.setDataset(topArtists);
        }
    }

    View.OnClickListener onArtistclickedListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            int position = artistsRecyclerView.getChildLayoutPosition(view);
            Artist artist = mAdapter.getItemByPosition(position);
            Intent intent = new Intent(getContext(), DetailsActivity.class);
            intent.putExtra("name", artist.getName());
            intent.putExtra("image", artist.getImageUrl());
            intent.putExtra("playCount", artist.getPlaycount());
            intent.putExtra("artist", "");
            intent.putExtra("duration", "");
            startActivity(intent);
        }
    };

    @Override
    public void showError() {
        Toast.makeText(getContext(), R.string.general_error, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void showEmpty() {
        emptyLayout.setVisibility(View.VISIBLE);
    }

    @Override
    public void hidEmpty() {
        emptyLayout.setVisibility(View.GONE);

    }

    public static TopArtistsFragment newInstance() {
        return new TopArtistsFragment();
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    public interface OnFragmentInteractionListener {

        void onArtistClicked(Artist artist);

    }
}

File: Artist/app/src/main/java/com/example/artist/ui/topartistslisting/TopArtistsFragment.kt
package com.example.artist.ui.topartistslisting

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.artist.R
import com.example.artist.adapters.TopArtistsAdapter
import com.example.artist.models.Artist
import com.example.artist.ui.BaseFragment
import com.example.artist.ui.DetailsActivity
import com.example.artist.ui.topartistslisting.di.DaggerTopArtistsComponent
import com.example.artist.ui.topartistslisting.di.TopArtistsModule
import com.example.artist.utils.Constants
import kotlinx.android.synthetic.main.fragment_artists.*
import javax.inject.Inject

class TopArtistsFragment : BaseFragment(), TopArtistsView {

    var mListener: OnFragmentInteractionListener? = null


    @Inject
    lateinit var mPresenter: TopArtistsPresenter
    var mAdapter: TopArtistsAdapter? = null
    override fun searchUserName(userName: String) {
        mAdapter?.clearDataset()
        mPresenter.getUserTopArtists(userName, Constants.API_KEY)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_artists, container, false)
        return view
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        DaggerTopArtistsComponent.builder().topArtistsModule(TopArtistsModule(this)).build().inject(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mPresenter.getUserTopArtists(Constants.DEFAULT_LASTFM_USER, Constants.API_KEY)
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        mListener = if (context is OnFragmentInteractionListener) {
            context
        } else {
            throw RuntimeException(context.toString()
                    + " must implement OnFragmentInteractionListener")
        }
    }

    override fun onDetach() {
        super.onDetach()
        mListener = null
    }

    override fun showProgress() {
        prgrs_main?.visibility = View.VISIBLE
    }

    override fun hideProgress() {
        prgrs_main?.visibility = View.GONE
    }

    override fun updateData(topArtists: MutableList<Artist>) {
        if (mAdapter == null) {
            mAdapter = context?.let { TopArtistsAdapter(topArtists, it, onArtistclickedListener) }
            val linearLayoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
            rclr_artists?.layoutManager = linearLayoutManager
            rclr_artists?.adapter = mAdapter
            mAdapter?.notifyDataSetChanged()
        } else {
            mAdapter?.setDataset(topArtists)
        }
    }

    var onArtistclickedListener = View.OnClickListener { view ->
        val position = rclr_artists?.getChildLayoutPosition(view)
        val artist = mAdapter?.getItemByPosition(position!!)
        val intent = Intent(context, DetailsActivity::class.java)
        intent.putExtra("name", artist?.name)
        intent.putExtra("image", artist?.imageUrl)
        intent.putExtra("playCount", artist?.playcount)
        intent.putExtra("artist", "")
        intent.putExtra("duration", "")
        startActivity(intent)
    }

    override fun showError() {
        Toast.makeText(context, R.string.general_error, Toast.LENGTH_SHORT).show()
    }

    override fun showEmpty() {
        empty_layout?.visibility = View.VISIBLE
    }

    override fun hidEmpty() {
        empty_layout?.visibility = View.GONE
    }

    /**
     * This interface must be implemented by activities that contain this
     * fragment to allow an interaction in this fragment to be communicated
     * to the activity and potentially other fragments contained in that
     * activity.
     */
    interface OnFragmentInteractionListener {
        fun onArtistClicked(artist: Artist)
    }

    companion object {
        fun newInstance(): TopArtistsFragment {
            return TopArtistsFragment()
        }
    }
}